File: companies/Meta/604.design-compressed-string-iterator.go
package leetcode

//
// @lc app=leetcode id=604 lang=golang
//
// [604] Design Compressed String Iterator
//
// Notes
// Keep the compressed string index, current character, and remaining repeat
// count. HasNext lazily loads the next letter+count group when the current group
// is exhausted; Next consumes one character or returns a space when no group
// remains. Time: amortized O(1) per operation. Space: O(1).
//
// @lc code=start

type StringIterator604 struct {
	compressed  string
	index       int
	currentChar byte
	remaining   int
}

func NewStringIterator604(compressedString string) *StringIterator604 {
	return &StringIterator604{compressed: compressedString}
}

func (s *StringIterator604) Next() byte {
	if !s.HasNext() {
		return ' '
	}
	s.remaining--
	return s.currentChar
}

func (s *StringIterator604) HasNext() bool {
	if s.remaining > 0 {
		return true
	}
	s.loadNextGroup()
	return s.remaining > 0
}

func (s *StringIterator604) loadNextGroup() {
	if s.index >= len(s.compressed) {
		return
	}

	s.currentChar = s.compressed[s.index]
	s.index++
	count := 0
	for s.index < len(s.compressed) && '0' <= s.compressed[s.index] && s.compressed[s.index] <= '9' {
		count = count*10 + int(s.compressed[s.index]-'0')
		s.index++
	}
	s.remaining = count
}

// @lc code=end

File: companies/Meta/604.design-compressed-string-iterator.java
/*
 * @lc app=leetcode id=604 lang=java
 *
 * [604] Design Compressed String Iterator
 *
 * Lazily parse one compressed group at a time. currentChar and remaining hold
 * the active run; when remaining reaches zero, parse the next letter and its
 * decimal count. This avoids expanding huge counts.
 *
 * Time: O(1) amortized per call. Space: O(1).
 */

// @lc code=start
class StringIterator {
    private final String compressed;
    private int index;
    private char currentChar;
    private int remaining;

    public StringIterator(String compressedString) {
        this.compressed = compressedString;
    }

    public char next() {
        if (!hasNext()) {
            return ' ';
        }
        remaining--;
        return currentChar;
    }

    public boolean hasNext() {
        if (remaining > 0) {
            return true;
        }
        loadNextGroup();
        return remaining > 0;
    }

    private void loadNextGroup() {
        if (index >= compressed.length()) {
            return;
        }
        currentChar = compressed.charAt(index++);
        int count = 0;
        while (index < compressed.length() && Character.isDigit(compressed.charAt(index))) {
            count = count * 10 + compressed.charAt(index) - '0';
            index++;
        }
        remaining = count;
    }
}
// @lc code=end
